### PR TITLE
Feat raspberry pi mike

### PR DESCRIPTION
**Pull request by MikeG3**
This Pull Request includes code that is installed on the Raspberry Pi

All of this code is also found in the branch "feat-Hardware_Communications-Mike", which contains all of the tags to the story notes that each commit relates too.

Unfortunately, this code was not added earlier, and due to merging conflicts, and complications of having code on two different systems, I am making a pull Request, that just contains the code on the Raspberry Pi

The code that contains backend scripting, to function with the Raspberry Pi, is still in the "feat-Hardware_Communications-Mike" branch, and a seperate Pull Request will be made for that code

### CODE REVIEW

File: RasberryPi/app.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    
+
+def sense_motion():
+    while True:
+        blue_led.off()
+        motion_detector.wait_for_motion() 
+        timestamp = time.strftime("%y%b%d%H%M%S")
+        print("Motion Detector Triggered " + timestamp)
+        try:
+            urllib.request.urlopen("http://169.254.135.232:8000/motion/" + timestamp)
```
Comment (hasan-ahmed) [inline RasberryPi/app.py:43]: I believe we should make the url a variable instead of hardcoding in. This will make the code more extensible & scaleable. It also seems like we are using this same url on line 46, so making it a variable would avoid code duplication as well.

`
url = "http://" + BackendServerIp +":"+ BackendServerPort + "/motion/" + timestamp
urllib.request.urlopen(url)
`

File: RasberryPi/app.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            print("Error sending motion detection timestamp")
+            print("http://" + BackendServerIp +":"+ BackendServerPort + "/motion/" + timestamp)
+        blue_led.on()
+        time.sleep(10.0)
+        blue_led.off()
+    
+
+@app.route('/')
+def index():
+    return render_template('index.html')
```
Comment (hasan-ahmed) [inline RasberryPi/app.py:54]: I am not sure if it makes sense to return an html file from the raspberry pi to a web app. We should look into returning json data that looks like this. But we should do this on another PR as it might be a breaking change with other components of the system.

`
{
success: true, 
message: "The door has been successfully opened"
}
`

File: RasberryPi/app.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+
+@app.route('/open.html')
+def open():
+    open_door()
+    return render_template('index.html')
```
Comment (hasan-ahmed) [inline RasberryPi/app.py:60]: I am not sure if it makes sense to return an html file from the raspberry pi to a web app. We should look into returning json data that looks like this. But we should do this on another PR as it might be a breaking change with other components of the system.

`
{
success: true, 
message: "The door has been successfully opened"
}
`

File: RasberryPi/app.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+
+@app.route('/submitbutton', methods=['POST'])
+def submitbutton():
+    open_door()
+    return render_template('index.html')
```
Comment (hasan-ahmed) [inline RasberryPi/app.py:66]: I am not sure if it makes sense to return an html file from the raspberry pi to a web app. We should look into returning json data that looks like this. But we should do this on another PR as it might be a breaking change with other components of the system.

`
{
success: true, 
message: "The door has been successfully opened"
}
`

**Review by hasan-ahmed — COMMENTED** [review]
**Variables Name**
Variable are well named with proper casing for constants vs. variables.

**Function Name**
Functions are well named

**Long Function**
I could not find any long functions

**Code Complexity**
Returning a full html file from the raspberry pi for every route is a little too complex and data intensive. This design would mean that we would need to create a new html file for every new route created, and that HTML file would never be shown to anyone. As mentioned in my inline review comments, we can instead return json data showing the success of operations in the response of routes. Here's a link that shows how to do that:
[https://riptutorial.com/flask/example/5831/return-a-json-response-from-flask-api](url)

PR Checklist

- [x] Documentation present for every component and function.
- [] Unit tests are implemented.
- [] Implementation matches the corresponding requirements.
- [] All commits linked to Jira ticket.

Overall Remarks
I know that the code here satisfies the requirements of the reqs that you worked on based on videos that have been shared in acceptance tests + Discord. But the commits and PR do not have any link to JIRA issues. I would recommend updating your PR description to include the Jira issue numbers so that eveyrthing can be linked properly. 

I have left some inline notes on the codes about places where I think the code can be improved, but you need to consider if it is best to make those changes in this PR, or in a separate future PR to avoid breaking changes. 

Overall, really good job and getting this up and running on a platform (Raspberry PI) that you have never used before.

**Comment by abirshah** [comment]
### _Variables Name._

Not a lot of variables for all the file changes.  HTML files are not really part of the logic. 
Only RaspberryPi/app.py  files _red_led_, _blue_led_  can explain more. which is like 
_red_led_pet_door_closed_
_blue_let_pet_door_open_

### _Function Name_

In l RaspberryPi/app.py
Line 27: _open_door_ can be improved to  _door_will_open_using_this_
Line 36: _sense_motion_ can be improved to _motion_sensor_will_tigger_


### _Long Function._

There are no long function in this PR


### _Code Complexity._

The build failed because of the bugs and security hotspots 

SonarCoud results are below.

Screen Shot 2022-03-31 at 11 22 35 PM


### _PR Checklist_

  - [x] Documentation present for every component and function.
  - [ ] Unit tests are implemented.
  - [x] Implementation matches the corresponding requirements.
  - [x] All commits linked to Jira ticket.
  
### _Overall Remarks._

Approved. Good work for taking care of RaspberryPi. Also thanks for creating ReadMe.txt file

**Review by abirshah — APPROVED** [review]

File: RasberryPi/app.py (hunk 4 above)
Comment (MikeG3) [reply RasberryPi/app.py:66]: Ok, I used that line of code at some point because I was getting an error without it. Ill try again. Im new to using Flask, Ill try your recommendation
